src/main.c

```diff
@@ -2223,17 +2279,40 @@ H2O_NORETURN static void *run_loop(void *_thread_index)
     /* setup listeners */
     for (i = 0; i != conf.num_listeners; ++i) {
         struct listener_config_t *listener_config = conf.listeners[i];
-        int fd;
+        int fd = -1;
         /* dup the listener fd for other threads than the main thread */
         if (thread_index == 0) {
             fd = listener_config->fd;
         } else {
-            if ((fd = dup(listener_config->fd)) == -1) {
+            int reuseport = 0;
+#if H2O_HTTP3_USE_REUSEPORT
+            socklen_t reuseportlen = sizeof(reuseport);
+            if (getsockopt(listener_config->fd, SOL_SOCKET, H2O_SO_REUSEPORT, &reuseport, &reuseportlen) != 0) {
+                perror("gestockopt(SO_REUSEPORT) failed");
+                abort();
+            }
+            assert(reuseportlen == sizeof(reuseport));
+            if (reuseport) {
+                struct sockaddr_storage ss;
+                socklen_t sslen = sizeof(ss);
+                if (getsockname(listener_config->fd, (struct sockaddr *)&ss, &sslen) != 0) {
+                    perror("failed to obtain local address of the listening QUIC socket");
+                    abort();
+                }
+                if ((fd = open_listener(ss.ss_family, SOCK_DGRAM, 0, (struct sockaddr *)&ss, sslen, 1)) != -1) {
+                    setsockopt_recvpktinfo(fd, ss.ss_family);
+                } else {
+                    reuseport = 0;
+                }
+            }
+#endif
+            if (!reuseport && (fd = dup(listener_config->fd)) == -1) {
```

Comment:
This is the crux adding support for SO_REUSEPORT.

Until now, we've use dup(2) so that every thread would have different file descriptors that map to the listening socket. Now, we check if the listening socket has the SO_REUSEPORT socket option set, and if so, create a new socket with the same configuration.